File: sdk-go/keys.go
package banhbaoring

import (
	"context"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// KeysService handles key management operations.
type KeysService struct {
	client *Client
}

// CreateKeyRequest is the request for creating a key.
type CreateKeyRequest struct {
	// Name is the key name (required).
	Name string `json:"name"`
	// NamespaceID is the namespace to create the key in (required).
	NamespaceID uuid.UUID `json:"namespace_id"`
	// Algorithm is the key algorithm (default: secp256k1).
	Algorithm string `json:"algorithm,omitempty"`
	// Exportable indicates if the private key can be exported.
	Exportable bool `json:"exportable,omitempty"`
	// Metadata is optional key-value metadata.
	Metadata map[string]string `json:"metadata,omitempty"`
}

// CreateBatchRequest creates multiple keys at once.
type CreateBatchRequest struct {
	// Prefix is the key name prefix. Keys will be named "prefix-1", "prefix-2", etc.
	Prefix string `json:"prefix"`
	// Count is the number of keys to create (1-100).
	Count int `json:"count"`
	// NamespaceID is the namespace to create keys in (required).
	NamespaceID uuid.UUID `json:"namespace_id"`
	// Exportable indicates if private keys can be exported.
	Exportable bool `json:"exportable,omitempty"`
}

// UpdateKeyRequest is the request for updating a key.
type UpdateKeyRequest struct {
	// Name is the new key name.
	Name string `json:"name,omitempty"`
	// Metadata is the new key metadata.
	Metadata map[string]string `json:"metadata,omitempty"`
}

// ImportKeyRequest is the request for importing a key.
type ImportKeyRequest struct {
	// Name is the key name (required).
	Name string `json:"name"`
	// NamespaceID is the namespace to import the key into (required).
	NamespaceID uuid.UUID `json:"namespace_id"`
	// PrivateKey is the base64-encoded private key (required).
	PrivateKey string `json:"private_key"`
	// Exportable indicates if the private key can be exported.
	Exportable bool `json:"exportable,omitempty"`
}

// ExportKeyResponse is the response from exporting a key.
type ExportKeyResponse struct {
	// PrivateKey is the base64-encoded private key.
	PrivateKey string `json:"private_key"`
	// Warning is a security warning message.
	Warning string `json:"warning"`
}

// keyResponse is the internal API response format for keys.
type keyResponse struct {
	ID          uuid.UUID              `json:"id"`
	NamespaceID uuid.UUID              `json:"namespace_id"`
	Name        string                 `json:"name"`
	PublicKey   string                 `json:"public_key"`
	Address     string                 `json:"address"`
	Algorithm   Algorithm              `json:"algorithm"`
	Exportable  bool                   `json:"exportable"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
	Version     int                    `json:"version"`
	CreatedAt   string                 `json:"created_at"`
}

// toKey converts a keyResponse to a Key.
func (r *keyResponse) toKey() *Key {
	createdAt, _ := time.Parse("2006-01-02T15:04:05Z", r.CreatedAt)

	metadata := make(map[string]string)
	for k, v := range r.Metadata {
		if s, ok := v.(string); ok {
			metadata[k] = s
		}
	}

	return &Key{
		ID:          r.ID,
		NamespaceID: r.NamespaceID,
		Name:        r.Name,
		PublicKey:   r.PublicKey,
		Address:     r.Address,
		Algorithm:   r.Algorithm,
		Exportable:  r.Exportable,
		Metadata:    metadata,
		Version:     r.Version,
		CreatedAt:   createdAt,
	}
}

// Create creates a new key.
//
// Example:
//
//	key, err := client.Keys.Create(ctx, banhbaoring.CreateKeyRequest{
//	    Name:        "sequencer-main",
//	    NamespaceID: namespaceID,
//	    Algorithm:   "secp256k1",
//	    Exportable:  true,
//	})
func (s *KeysService) Create(ctx context.Context, req CreateKeyRequest) (*Key, error) {
	// Convert to API format
	apiReq := map[string]interface{}{
		"name":         req.Name,
		"namespace_id": req.NamespaceID.String(),
		"exportable":   req.Exportable,
	}
	if req.Algorithm != "" {
		apiReq["algorithm"] = req.Algorithm
	}
	if len(req.Metadata) > 0 {
		apiReq["metadata"] = req.Metadata
	}

	var resp keyResponse
	if err := s.client.post(ctx, "/v1/keys", apiReq, &resp); err != nil {
		return nil, err
	}
	return resp.toKey(), nil
}

// CreateBatch creates multiple keys in parallel.
// This is optimized for Celestia's parallel worker pattern.
//
// Example:
//
//	keys, err := client.Keys.CreateBatch(ctx, banhbaoring.CreateBatchRequest{
//	    Prefix:      "blob-worker",
//	    Count:       4,
//	    NamespaceID: prodNamespace,
//	})
//	// Creates: blob-worker-1, blob-worker-2, blob-worker-3, blob-worker-4
func (s *KeysService) CreateBatch(ctx context.Context, req CreateBatchRequest) ([]*Key, error) {
	apiReq := map[string]interface{}{
		"prefix":       req.Prefix,
		"count":        req.Count,
		"namespace_id": req.NamespaceID.String(),
		"exportable":   req.Exportable,
	}

	var resp struct {
		Keys  []*keyResponse `json:"keys"`
		Count int            `json:"count"`
	}
	if err := s.client.post(ctx, "/v1/keys/batch", apiReq, &resp); err != nil {
		return nil, err
	}

	keys := make([]*Key, len(resp.Keys))
	for i, k := range resp.Keys {
		keys[i] = k.toKey()
	}
	return keys, nil
}

// Get retrieves a key by ID.
//
// Example:
//
//	key, err := client.Keys.Get(ctx, keyID)
func (s *KeysService) Get(ctx context.Context, keyID uuid.UUID) (*Key, error) {
	var resp keyResponse
	if err := s.client.get(ctx, fmt.Sprintf("/v1/keys/%s", keyID), &resp); err != nil {
		return nil, err
	}
	return resp.toKey(), nil
}

// ListOptions are options for listing keys.
type ListOptions struct {
	// NamespaceID filters keys by namespace.
	NamespaceID *uuid.UUID
}

// List returns all keys, optionally filtered by namespace.
//
// Example:
//
//	// List all keys
//	keys, err := client.Keys.List(ctx, nil)
//
//	// List keys in a specific namespace
//	keys, err := client.Keys.List(ctx, &banhbaoring.ListOptions{
//	    NamespaceID: &namespaceID,
//	})
func (s *KeysService) List(ctx context.Context, opts *ListOptions) ([]*Key, error) {
	path := "/v1/keys"
	if opts != nil && opts.NamespaceID != nil {
		path = fmt.Sprintf("/v1/keys?namespace_id=%s", opts.NamespaceID)
	}

	var resp []*keyResponse
	if err := s.client.get(ctx, path, &resp); err != nil {
		return nil, err
	}

	keys := make([]*Key, len(resp))
	for i, k := range resp {
		keys[i] = k.toKey()
	}
	return keys, nil
}

// Delete deletes a key.
//
// Example:
//
//	err := client.Keys.Delete(ctx, keyID)
func (s *KeysService) Delete(ctx context.Context, keyID uuid.UUID) error {
	return s.client.delete(ctx, fmt.Sprintf("/v1/keys/%s", keyID))
}

// Import imports a private key.
//
// Example:
//
//	key, err := client.Keys.Import(ctx, banhbaoring.ImportKeyRequest{
//	    Name:        "imported-key",
//	    NamespaceID: namespaceID,
//	    PrivateKey:  base64PrivateKey,
//	    Exportable:  true,
//	})
func (s *KeysService) Import(ctx context.Context, req ImportKeyRequest) (*Key, error) {
	apiReq := map[string]interface{}{
		"name":         req.Name,
		"namespace_id": req.NamespaceID.String(),
		"private_key":  req.PrivateKey,
		"exportable":   req.Exportable,
	}

	var resp keyResponse
	if err := s.client.post(ctx, "/v1/keys/import", apiReq, &resp); err != nil {
		return nil, err
	}
	return resp.toKey(), nil
}

// Export exports a key's private key material.
// The key must have been created with Exportable: true.
//
// Example:
//
//	result, err := client.Keys.Export(ctx, keyID)
//	privateKey := result.PrivateKey // base64-encoded
func (s *KeysService) Export(ctx context.Context, keyID uuid.UUID) (*ExportKeyResponse, error) {
	var resp ExportKeyResponse
	if err := s.client.post(ctx, fmt.Sprintf("/v1/keys/%s/export", keyID), nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

File: sdk-rust/src/keys.rs
//! Key management operations.
//!
//! This module provides the KeysClient for creating, retrieving, listing,
//! and deleting cryptographic keys.

use crate::client::Client;
use crate::error::Result;
use crate::types::{CreateBatchRequest, CreateKeyRequest, Key};
use serde::Deserialize;
use uuid::Uuid;

/// Client for key management operations.
///
/// Access via `client.keys()`.
pub struct KeysClient {
    client: Client,
}

impl KeysClient {
    pub(crate) fn new(client: Client) -> Self {
        Self { client }
    }

    /// Create a new key.
    ///
    /// # Arguments
    ///
    /// * `request` - Key creation parameters
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// use banhbaoring::{Client, types::CreateKeyRequest};
    /// use uuid::Uuid;
    ///
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     let client = Client::new("bbr_live_xxxxx");
    ///     let namespace_id = Uuid::parse_str("...")?;
    ///     
    ///     let key = client.keys().create(CreateKeyRequest {
    ///         name: "my-key".to_string(),
    ///         namespace_id,
    ///         algorithm: Some("secp256k1".to_string()),
    ///         exportable: Some(false),
    ///         metadata: None,
    ///     }).await?;
    ///     
    ///     println!("Created key: {} ({})", key.name, key.address);
    ///     Ok(())
    /// }
    /// ```
    pub async fn create(&self, request: CreateKeyRequest) -> Result<Key> {
        self.client.post("/v1/keys", &request).await
    }

    /// Create multiple keys in parallel.
    ///
    /// Optimized for Celestia's parallel worker pattern. Creates keys with
    /// sequential names like "prefix-1", "prefix-2", etc.
    ///
    /// # Arguments
    ///
    /// * `request` - Batch creation parameters
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// use banhbaoring::{Client, types::CreateBatchRequest};
    /// use uuid::Uuid;
    ///
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     let client = Client::new("bbr_live_xxxxx");
    ///     let namespace_id = Uuid::parse_str("...")?;
    ///     
    ///     let keys = client.keys().create_batch(CreateBatchRequest {
    ///         prefix: "blob-worker".to_string(),
    ///         count: 4,
    ///         namespace_id,
    ///         exportable: None,
    ///     }).await?;
    ///     
    ///     // Creates: blob-worker-1, blob-worker-2, blob-worker-3, blob-worker-4
    ///     for key in &keys {
    ///         println!("  {}: {}", key.name, key.address);
    ///     }
    ///     Ok(())
    /// }
    /// ```
    pub async fn create_batch(&self, request: CreateBatchRequest) -> Result<Vec<Key>> {
        #[derive(Deserialize)]
        struct Response {
            keys: Vec<Key>,
        }

        let response: Response = self.client.post("/v1/keys/batch", &request).await?;
        Ok(response.keys)
    }

    /// Get a key by ID.
    ///
    /// # Arguments
    ///
    /// * `key_id` - The key's unique identifier
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// use banhbaoring::Client;
    /// use uuid::Uuid;
    ///
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     let client = Client::new("bbr_live_xxxxx");
    ///     let key_id = Uuid::parse_str("...")?;
    ///     
    ///     let key = client.keys().get(&key_id).await?;
    ///     println!("Key: {} ({})", key.name, key.algorithm);
    ///     Ok(())
    /// }
    /// ```
    pub async fn get(&self, key_id: &Uuid) -> Result<Key> {
        self.client.get(&format!("/v1/keys/{}", key_id)).await
    }

    /// List all keys, optionally filtered by namespace.
    ///
    /// # Arguments
    ///
    /// * `namespace_id` - Optional namespace ID to filter by
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// use banhbaoring::Client;
    /// use uuid::Uuid;
    ///
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     let client = Client::new("bbr_live_xxxxx");
    ///     
    ///     // List all keys
    ///     let all_keys = client.keys().list(None).await?;
    ///     println!("Total keys: {}", all_keys.len());
    ///     
    ///     // List keys in a specific namespace
    ///     let namespace_id = Uuid::parse_str("...")?;
    ///     let namespace_keys = client.keys().list(Some(&namespace_id)).await?;
    ///     Ok(())
    /// }
    /// ```
    pub async fn list(&self, namespace_id: Option<&Uuid>) -> Result<Vec<Key>> {
        let path = match namespace_id {
            Some(id) => format!("/v1/keys?namespace_id={}", id),
            None => "/v1/keys".to_string(),
        };
        self.client.get(&path).await
    }

    /// Delete a key.
    ///
    /// **Warning:** This operation is irreversible. The key and all associated
    /// data will be permanently deleted.
    ///
    /// # Arguments
    ///
    /// * `key_id` - The key's unique identifier
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// use banhbaoring::Client;
    /// use uuid::Uuid;
    ///
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     let client = Client::new("bbr_live_xxxxx");
    ///     let key_id = Uuid::parse_str("...")?;
    ///     
    ///     client.keys().delete(&key_id).await?;
    ///     println!("Key deleted");
    ///     Ok(())
    /// }
    /// ```
    pub async fn delete(&self, key_id: &Uuid) -> Result<()> {
        self.client.delete(&format!("/v1/keys/{}", key_id)).await
    }

    /// Get a key by name within a namespace.
    ///
    /// # Arguments
    ///
    /// * `namespace_id` - The namespace ID
    /// * `name` - The key name
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// use banhbaoring::Client;
    /// use uuid::Uuid;
    ///
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     let client = Client::new("bbr_live_xxxxx");
    ///     let namespace_id = Uuid::parse_str("...")?;
    ///     
    ///     let key = client.keys().get_by_name(&namespace_id, "my-key").await?;
    ///     println!("Key ID: {}", key.id);
    ///     Ok(())
    /// }
    /// ```
    pub async fn get_by_name(&self, namespace_id: &Uuid, name: &str) -> Result<Key> {
        self.client
            .get(&format!("/v1/keys/by-name/{}/{}", namespace_id, name))
            .await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_keys_client_creation() {
        let client = Client::new("test_key");
        let _keys = client.keys();
        // Just verify it compiles and doesn't panic
    }
}

